File: rusty-js/modules/http/src/request.rs
use http_crate::{Method, Uri};
use rusty_js::{function::Optional, *};

use crate::{blob::Blob, header::Headers};

#[js_class]
pub struct Request {
    method: Method,
    url: Uri,
    url_string: String, // Store the original URL string
    headers: Headers,
    body: Option<Blob>,
    mode: RequestMode,
    credentials: RequestCredentials,
    cache: RequestCache,
    redirect: RequestRedirect,
    referrer: String,
    referrer_policy: ReferrerPolicy,
    integrity: String,
    keepalive: bool,
    is_reload_navigation: bool,
    is_history_navigation: bool,
    signal: Option<JSObject>, // AbortSignal
}

#[derive(Default, Clone)]
enum RequestMode {
    #[default]
    Cors,
    NoCors,
    SameOrigin,
    Navigate,
}

#[derive(Default, Clone)]
enum RequestCredentials {
    #[default]
    SameOrigin,
    Omit,
    Include,
}

#[derive(Default, Clone)]
enum RequestCache {
    #[default]
    Default,
    NoStore,
    Reload,
    NoCache,
    ForceCache,
    OnlyIfCached,
}

#[derive(Default, Clone)]
enum RequestRedirect {
    #[default]
    Follow,
    Error,
    Manual,
}

#[derive(Default, Clone)]
enum ReferrerPolicy {
    #[default]
    NoReferrer,
    NoReferrerWhenDowngrade,
    Origin,
    OriginWhenCrossOrigin,
    SameOrigin,
    StrictOrigin,
    StrictOriginWhenCrossOrigin,
    UnsafeUrl,
}

#[js_methods]
impl Request {
    fn is_valid_method(method: &str) -> bool {
        matches!(
            method.to_uppercase().as_str(),
            "GET" | "POST" | "PUT" | "DELETE" | "HEAD" | "OPTIONS" | "CONNECT" | "TRACE" | "PATCH"
        )
    }

    #[js_method(constructor)]
    pub fn new(input: JSValue, init: Optional<JSObject>) -> JSResult<Self> {
        println!("Creating new Request with input: {:?}", input);
        // Parse input - can be a URL string or another Request object
        let (method, url, url_string, headers, body) = if let Some(obj) =
            input.clone().into_object()
        {
            if let Ok(request) = obj.borrow::<Request>() {
                // Input is a Request object
                (
                    request.method.clone(),
                    request.url.clone(),
                    request.url_string.clone(),
                    request.headers.clone(),
                    request.body.clone(),
                )
            } else {
                // Input should be a URL string
                let url_str: String = input.try_into()?;
                // Validate URL format
                if !url_str.starts_with("http://") && !url_str.starts_with("https://") {
                    return Err(RustyJSError::TypeError(format!("Invalid URL: {}", url_str)));
                }
                let url = Uri::try_from(url_str.as_str())
                    .map_err(|_| RustyJSError::TypeError(format!("Invalid URL: {}", url_str)))?;
                (Method::GET, url, url_str, Headers::default(), None)
            }
        } else {
            // Input should be a URL string
            let url_str: String = input.try_into()?;
            // Validate URL format
            if !url_str.starts_with("http://") && !url_str.starts_with("https://") {
                return Err(RustyJSError::TypeError(format!("Invalid URL: {}", url_str)));
            }
            let url = Uri::try_from(url_str.as_str())
                .map_err(|_| RustyJSError::TypeError(format!("Invalid URL: {}", url_str)))?;
            (Method::GET, url, url_str, Headers::default(), None)
        };

        let mut request = Self {
            method,
            url,
            url_string,
            headers,
            body,
            mode: RequestMode::default(),
            credentials: RequestCredentials::default(),
            cache: RequestCache::default(),
            redirect: RequestRedirect::default(),
            referrer: String::new(),
            referrer_policy: ReferrerPolicy::default(),
            integrity: String::new(),
            keepalive: false,
            is_reload_navigation: false,
            is_history_navigation: false,
            signal: None,
        };

        // Process init object if provided
        if let Some(init) = init.0 {
            // Method
            if let Ok(method_str) = init.get::<_, String>("method") {
                if !Self::is_valid_method(&method_str) {
                    return Err(RustyJSError::TypeError(format!(
                        "Invalid method: {}",
                        method_str
                    )));
                }
                request.method = Method::from_bytes(method_str.as_bytes()).map_err(|_| {
                    RustyJSError::TypeError(format!("Invalid method: {}", method_str))
                })?;
            }

            // Headers
            if let Ok(headers_init) = init.get::<_, JSValue>("headers") {
                request.headers = Headers::new(Optional(Some(headers_init)))?;
            }

            // Body
            if let Ok(body_value) = init.get::<_, JSValue>("body") {
                if let Some(obj) = body_value.into_object() {
                    if let Ok(blob) = obj.borrow::<Blob>() {
                        request.body = Some(blob.clone());
                    }
                }
            }

            // Mode
            if let Ok(mode_str) = init.get::<_, String>("mode") {
                request.mode = match mode_str.as_str() {
                    "cors" => RequestMode::Cors,
                    "no-cors" => RequestMode::NoCors,
                    "same-origin" => RequestMode::SameOrigin,
                    "navigate" => RequestMode::Navigate,
                    _ => {
                        return Err(RustyJSError::TypeError(format!(
                            "Invalid mode: {}",
                            mode_str
                        )))
                    }
                };
            }

            // Credentials
            if let Ok(cred_str) = init.get::<_, String>("credentials") {
                request.credentials = match cred_str.as_str() {
                    "omit" => RequestCredentials::Omit,
                    "same-origin" => RequestCredentials::SameOrigin,
                    "include" => RequestCredentials::Include,
                    _ => {
                        return Err(RustyJSError::TypeError(format!(
                            "Invalid credentials: {}",
                            cred_str
                        )))
                    }
                };
            }

            // Cache
            if let Ok(cache_str) = init.get::<_, String>("cache") {
                request.cache = match cache_str.as_str() {
                    "default" => RequestCache::Default,
                    "no-store" => RequestCache::NoStore,
                    "reload" => RequestCache::Reload,
                    "no-cache" => RequestCache::NoCache,
                    "force-cache" => RequestCache::ForceCache,
                    "only-if-cached" => RequestCache::OnlyIfCached,
                    _ => {
                        return Err(RustyJSError::TypeError(format!(
                            "Invalid cache: {}",
                            cache_str
                        )))
                    }
                };
            }

            // Redirect
            if let Ok(redirect_str) = init.get::<_, String>("redirect") {
                request.redirect = match redirect_str.as_str() {
                    "follow" => RequestRedirect::Follow,
                    "error" => RequestRedirect::Error,
                    "manual" => RequestRedirect::Manual,
                    _ => {
                        return Err(RustyJSError::TypeError(format!(
                            "Invalid redirect: {}",
                            redirect_str
                        )))
                    }
                };
            }

            // Referrer
            if let Ok(referrer) = init.get::<_, String>("referrer") {
                request.referrer = referrer;
            }

            // ReferrerPolicy
            if let Ok(policy) = init.get::<_, String>("referrerPolicy") {
                request.referrer_policy = match policy.as_str() {
                    "no-referrer" => ReferrerPolicy::NoReferrer,
                    "no-referrer-when-downgrade" => ReferrerPolicy::NoReferrerWhenDowngrade,
                    "origin" => ReferrerPolicy::Origin,
                    "origin-when-cross-origin" => ReferrerPolicy::OriginWhenCrossOrigin,
                    "same-origin" => ReferrerPolicy::SameOrigin,
                    "strict-origin" => ReferrerPolicy::StrictOrigin,
                    "strict-origin-when-cross-origin" => {
                        ReferrerPolicy::StrictOriginWhenCrossOrigin
                    }
                    "unsafe-url" => ReferrerPolicy::UnsafeUrl,
                    _ => {
                        return Err(RustyJSError::TypeError(format!(
                            "Invalid referrerPolicy: {}",
                            policy
                        )))
                    }
                };
            }

            // Integrity
            if let Ok(integrity) = init.get::<_, String>("integrity") {
                request.integrity = integrity;
            }

            // Keepalive
            if let Ok(keepalive) = init.get::<_, bool>("keepalive") {
                request.keepalive = keepalive;
            }

            // Signal
            if let Ok(signal) = init.get::<_, JSObject>("signal") {
                request.signal = Some(signal);
            }
        }

        Ok(request)
    }

    #[js_method(getter)]
    pub fn method(&self) -> String {
        self.method.as_str().to_string()
    }

    #[js_method(getter)]
    pub fn url(&self) -> String {
        self.url_string.clone()
    }

    #[js_method(getter)]
    pub fn headers(&self) -> Headers {
        self.headers.clone()
    }

    #[js_method(getter)]
    pub fn destination(&self) -> String {
        // TODO: Implement destination getter
        "".to_string()
    }

    #[js_method(getter)]
    pub fn referrer(&self) -> String {
        self.referrer.clone()
    }

    #[js_method(getter, rename = "referrerPolicy")]
    pub fn referrer_policy(&self) -> String {
        match self.referrer_policy {
            ReferrerPolicy::NoReferrer => "no-referrer",
            ReferrerPolicy::NoReferrerWhenDowngrade => "no-referrer-when-downgrade",
            ReferrerPolicy::Origin => "origin",
            ReferrerPolicy::OriginWhenCrossOrigin => "origin-when-cross-origin",
            ReferrerPolicy::SameOrigin => "same-origin",
            ReferrerPolicy::StrictOrigin => "strict-origin",
            ReferrerPolicy::StrictOriginWhenCrossOrigin => "strict-origin-when-cross-origin",
            ReferrerPolicy::UnsafeUrl => "unsafe-url",
        }
        .to_string()
    }

    #[js_method(getter)]
    pub fn mode(&self) -> String {
        match self.mode {
            RequestMode::Cors => "cors",
            RequestMode::NoCors => "no-cors",
            RequestMode::SameOrigin => "same-origin",
            RequestMode::Navigate => "navigate",
        }
        .to_string()
    }

    #[js_method(getter)]
    pub fn credentials(&self) -> String {
        match self.credentials {
            RequestCredentials::Omit => "omit",
            RequestCredentials::SameOrigin => "same-origin",
            RequestCredentials::Include => "include",
        }
        .to_string()
    }

    #[js_method(getter)]
    pub fn cache(&self) -> String {
        match self.cache {
            RequestCache::Default => "default",
            RequestCache::NoStore => "no-store",
            RequestCache::Reload => "reload",
            RequestCache::NoCache => "no-cache",
            RequestCache::ForceCache => "force-cache",
            RequestCache::OnlyIfCached => "only-if-cached",
        }
        .to_string()
    }

    #[js_method(getter)]
    pub fn redirect(&self) -> String {
        match self.redirect {
            RequestRedirect::Follow => "follow",
            RequestRedirect::Error => "error",
            RequestRedirect::Manual => "manual",
        }
        .to_string()
    }

    #[js_method(getter)]
    pub fn integrity(&self) -> String {
        self.integrity.clone()
    }

    #[js_method(getter)]
    pub fn keepalive(&self) -> bool {
        self.keepalive
    }

    #[js_method(getter, rename = "isReloadNavigation")]
    pub fn is_reload_navigation(&self) -> bool {
        self.is_reload_navigation
    }

    #[js_method(getter, rename = "isHistoryNavigation")]
    pub fn is_history_navigation(&self) -> bool {
        self.is_history_navigation
    }

    #[js_method(getter)]
    pub fn signal(&self) -> Option<JSObject> {
        self.signal.clone()
    }

    #[js_method(getter)]
    pub fn body(&self) -> Option<Blob> {
        self.body.clone()
    }

    #[js_method]
    pub fn clone(&self) -> Self {
        Self {
            method: self.method.clone(),
            url: self.url.clone(),
            url_string: self.url_string.clone(),
            headers: self.headers.clone(),
            body: self.body.clone(),
            mode: self.mode.clone(),
            credentials: self.credentials.clone(),
            cache: self.cache.clone(),
            redirect: self.redirect.clone(),
            referrer: self.referrer.clone(),
            referrer_policy: self.referrer_policy.clone(),
            integrity: self.integrity.clone(),
            keepalive: self.keepalive,
            is_reload_navigation: self.is_reload_navigation,
            is_history_navigation: self.is_history_navigation,
            signal: self.signal.clone(),
        }
    }

    #[js_method]
    pub async fn text(&self) -> JSResult<String> {
        if let Some(blob) = &self.body {
            blob.text().await
        } else {
            Ok(String::new())
        }
    }

    #[js_method]
    pub async fn json(&self, ctx: JSContext) -> JSResult<JSValue> {
        let text = self.text().await?;
        ctx.eval(Source::from_bytes(text.as_bytes()))
    }

    #[js_method(rename = "arrayBuffer")]
    pub async fn array_buffer(&self, ctx: JSContext) -> JSResult<JSArrayBuffer<u8>> {
        if let Some(blob) = &self.body {
            blob.array_buffer(ctx).await
        } else {
            JSArrayBuffer::from_bytes(&ctx, &[])
        }
    }

    #[js_method(rename = "formData")]
    pub async fn form_data(&self) -> JSResult<JSObject> {
        // TODO: Implement form data parsing
        Err(RustyJSError::TypeError("Not implemented".to_string()))
    }
}

impl Default for Request {
    fn default() -> Self {
        let url_str = "about:blank".to_string();
        Self {
            method: Method::GET,
            url: Uri::from_static("about:blank"),
            url_string: url_str,
            headers: Headers::default(),
            body: None,
            mode: RequestMode::default(),
            credentials: RequestCredentials::default(),
            cache: RequestCache::default(),
            redirect: RequestRedirect::default(),
            referrer: String::new(),
            referrer_policy: ReferrerPolicy::default(),
            integrity: String::new(),
            keepalive: false,
            is_reload_navigation: false,
            is_history_navigation: false,
            signal: None,
        }
    }
}

pub(crate) fn init(ctx: &JSContext) -> JSResult<()> {
    ctx.register_class::<Request>();
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use rustyjs_test::*;

    #[test]
    fn test_request() {
        async_run!(|ctx: JSContext| async move {
            ctx.global().set(
                "print",
                JSFunc::new(&ctx, |msg: String| println!("JS: {}", msg)),
            );

            // Register console.log for better test output
            let source = Source::from_bytes(
                r#"
                const console = {
                    log: function(...args) {
                        print(args.join(' '))
                    }
                }
                "#,
            );
            let _ = ctx.eval::<()>(source);

            ctx.eval::<()>(Source::from_bytes(
                r#"
                class TextDecoder {
                    decode(arr) {
                        let str = '';
                        const len = arr.length;
                        let i = 0;

                        while (i < len) {
                            let charCode;

                            if (arr[i] < 0x80) {
                                charCode = arr[i++];
                            } else if (arr[i] < 0xE0) {
                                charCode = ((arr[i++] & 0x1F) << 6) |
                                         (arr[i++] & 0x3F);
                            } else {
                                charCode = ((arr[i++] & 0x0F) << 12) |
                                         ((arr[i++] & 0x3F) << 6) |
                                         (arr[i++] & 0x3F);
                            }

                            str += String.fromCharCode(charCode);
                        }

                        return str;
                    }
                }
                "#,
            ))
            .unwrap();

            // Initialize Blob first
            crate::blob::init(&ctx).unwrap();
            // Then initialize Headers
            crate::header::init(&ctx).unwrap();
            // Finally initialize Request
            init(&ctx).unwrap();

            let source = Source::from_path("tests/request.js").await.unwrap();
            let obj: JSObject = ctx.eval_async(source).await?;

            let total: i32 = obj.get("total").unwrap();
            let passed: i32 = obj.get("passed").unwrap();
            let success: bool = obj.get("success").unwrap();

            if !success {
                let failed: JSArray = obj.get("failed").unwrap();
                let error_messages: Vec<String> = failed.iter().collect::<JSResult<_>>()?;
                panic!(
                    "Request tests failed:\nPassed {}/{}\nFailures:\n{}",
                    passed,
                    total,
                    error_messages.join("\n")
                );
            }
            Ok(())
        });
    }
}

File: rusty-js/modules/http/tests/request.js
// Simple assertion functions
function assert(condition, message) {
  if (!condition) {
    print("Assertion failed: " + message);
    throw message || "Assertion failed";
  }
  return true;
}

function assertEqual(actual, expected, message) {
  if (actual != expected) {
    let error = `Expected value "${expected}" but got "${actual}"${message ? ": " + message : ""}`;
    print("Assertion failed: " + error);
    throw error;
  }
  return true;
}

// Test results storage
const results = {
  total: 0,
  passed: 0,
  failed: [],
};

// Run tests and return results
(async function () {
  try {
    // Test constructor with URL string
    results.total += 3;
    print("Testing constructor with URL string...");
    const request = new Request("https://example.com");
    results.passed += assert(
      request.method === "GET",
      "Default method should be GET",
    );
    results.passed += assert(
      request.url === "https://example.com",
      "URL should be set correctly",
    );
    results.passed += assert(
      request.headers instanceof Headers,
      "Headers should be instance of Headers",
    );

    // Test constructor with init options
    results.total += 5;
    print("Testing constructor with init options...");
    const requestWithInit = new Request("https://example.com", {
      method: "POST",
      mode: "no-cors",
      credentials: "include",
      cache: "no-cache",
      redirect: "error",
      headers: {
        "Content-Type": "application/json",
      },
    });

    results.passed += assert(
      requestWithInit.method === "POST",
      "Method should be set from init",
    );
    results.passed += assert(
      requestWithInit.mode === "no-cors",
      "Mode should be set from init",
    );
    results.passed += assert(
      requestWithInit.credentials === "include",
      "Credentials should be set from init",
    );
    results.passed += assert(
      requestWithInit.cache === "no-cache",
      "Cache should be set from init",
    );
    results.passed += assert(
      requestWithInit.headers.get("content-type") === "application/json",
      "Headers should be set from init",
    );

    // Test constructor with Request object
    results.total += 3;
    print("Testing constructor with Request object...");
    const original = new Request("https://example.com", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
    });
    const copy = new Request(original);

    results.passed += assert(
      copy.method === original.method,
      "Method should be copied",
    );
    results.passed += assert(copy.url === original.url, "URL should be copied");
    results.passed += assert(
      copy.headers.get("content-type") === original.headers.get("content-type"),
      "Headers should be copied",
    );

    // Test clone method
    results.total += 3;
    print("Testing clone method...");
    const cloned = original.clone();
    results.passed += assert(
      cloned.method === original.method,
      "Cloned method should match",
    );
    results.passed += assert(
      cloned.url === original.url,
      "Cloned URL should match",
    );
    results.passed += assert(
      cloned.headers.get("content-type") ===
        original.headers.get("content-type"),
      "Cloned headers should match",
    );

    // Test invalid inputs
    results.total += 6;
    print("Testing invalid inputs...");
    let hasError = false;

    try {
      new Request("not-a-url");
    } catch (e) {
      hasError = true;
      results.passed += assert(
        e instanceof TypeError,
        "Invalid URL should throw TypeError",
      );
    }
    if (!hasError) throw "Should throw error for invalid URL";

    hasError = false;
    try {
      new Request("https://example.com", { method: "INVALID" });
    } catch (e) {
      hasError = true;
      results.passed += assert(
        e instanceof TypeError,
        "Invalid method should throw TypeError",
      );
    }
    if (!hasError) throw "Should throw error for invalid method";

    // Test other invalid options
    ["mode", "credentials", "cache", "redirect"].forEach((option) => {
      hasError = false;
      try {
        const init = {};
        init[option] = "invalid-value";
        new Request("https://example.com", init);
      } catch (e) {
        hasError = true;
        results.passed += assert(
          e instanceof TypeError,
          `Invalid ${option} should throw TypeError`,
        );
      }
      if (!hasError) throw `Should throw error for invalid ${option}`;
    });

    // Test getters
    results.total += 13;
    print("Testing getters...");
    const requestWithAllProps = new Request("https://example.com", {
      method: "POST",
      mode: "no-cors",
      credentials: "include",
      cache: "no-cache",
      redirect: "error",
      referrer: "https://example.com/referrer",
      referrerPolicy: "strict-origin",
      integrity: "sha256-hash",
      keepalive: true,
      headers: {
        "Content-Type": "application/json",
      },
    });

    results.passed += assert(
      requestWithAllProps.destination === "",
      "Destination should be empty string",
    );
    results.passed += assert(
      requestWithAllProps.referrer === "https://example.com/referrer",
      "Referrer should match",
    );
    results.passed += assert(
      requestWithAllProps.referrerPolicy === "strict-origin",
      "ReferrerPolicy should match",
    );
    results.passed += assert(
      requestWithAllProps.mode === "no-cors",
      "Mode should match",
    );
    results.passed += assert(
      requestWithAllProps.credentials === "include",
      "Credentials should match",
    );
    results.passed += assert(
      requestWithAllProps.cache === "no-cache",
      "Cache should match",
    );
    results.passed += assert(
      requestWithAllProps.redirect === "error",
      "Redirect should match",
    );
    results.passed += assert(
      requestWithAllProps.integrity === "sha256-hash",
      "Integrity should match",
    );
    results.passed += assert(
      requestWithAllProps.keepalive === true,
      "Keepalive should match",
    );
    results.passed += assert(
      requestWithAllProps.isReloadNavigation === false,
      "isReloadNavigation should be false by default",
    );
    results.passed += assert(
      requestWithAllProps.isHistoryNavigation === false,
      "isHistoryNavigation should be false by default",
    );
    results.passed += assert(
      requestWithAllProps.signal === null,
      "Signal should be null by default",
    );
    results.passed += assert(
      requestWithAllProps.body === null,
      "Body should be null by default",
    );

    // Test async methods
    results.total += 2;
    print("Testing async methods...");

    // Create a request with text body
    const textBlob = new Blob(["Hello World"], { type: "text/plain" });
    const requestWithBody = new Request("https://example.com", {
      method: "POST",
      body: textBlob,
    });

    // Test text() method
    const text = await requestWithBody.text();
    results.passed += assert(
      text === "Hello World",
      "text() should return correct content",
    );

    // Test json() method
    // const jsonBlob = new Blob(['{"message":"Hello"}'], {
    //   type: "application/json",
    // });
    // const jsonRequest = new Request("https://example.com", {
    //   method: "POST",
    //   body: jsonBlob,
    // });
    // const json = await jsonRequest.json();
    // results.passed += assert(
    //   json.message === "Hello",
    //   "json() should parse JSON correctly",
    // );

    // Test arrayBuffer() method
    const buffer = await requestWithBody.arrayBuffer();
    const view = new Uint8Array(buffer);
    const decoder = new TextDecoder();
    results.passed += assert(
      decoder.decode(view) === "Hello World",
      "arrayBuffer() should return correct content",
    );

    // Test formData() method - should throw not implemented error
    // let hasFormDataError = false;
    // try {
    //   await requestWithBody.formData();
    // } catch (e) {
    //   hasFormDataError =
    //     e instanceof TypeError && e.message.includes("Not implemented");
    //   results.passed += hasFormDataError;
    // }
    // if (!hasFormDataError)
    //   throw "formData() should throw not implemented error";

    print(`Tests completed: ${results.passed}/${results.total} passed`);
  } catch (err) {
    print("Test failed with error: " + err);
    print(
      "Error details:",
      err instanceof Error ? err.stack : "No stack trace available",
    );
    results.failed.push(err.toString());
  }

  return {
    total: results.total,
    passed: results.passed,
    failed: results.failed,
    success: results.failed.length === 0 && results.total === results.passed,
  };
})();
